zkvm/docs/zkvm-impl.md

```diff
@@ -0,0 +1,75 @@
+`VM` executes the instructions, producing a serialized program bytecode, aggregated transaction signature
+and a zero-knowledge R1CS proof for the [Cloak protocol](../../spacesuit/spec.md) and custom statements expressed by instructions.
+
+`Verifier` is an API for _verifying_ a transaction object `Tx`: it parses the bytecode, executes it, verifies the aggregated transaction signature and the R1CS proof, producing a `VerifiedTx` as a result. Verification logic is described by the [ZkVM specification](zkvm-spec.md).
+
+Note that `VM` execution is used twice: first, for _proving_, then for _verifying_. 
+
+How does the `Prover` know how to sign transaction and make a proof? The prover’s input is not an opaque sequence of instruction codes, but _witness-bearing instructions_. That is, a `push` instruction on the prover’s side does not hold an opaque string of bytes, but an accurate _witness type_ that may contain secret data and necessary structure for creating the proofs and signatures.
+
+## Opaque and witness types
+
+We call a type **opaque** if it provides only enough information for the _verification_ of a ZkVM transaction or some sub-protocol.
+
+We call a type **witness** if it contains secret data and structure necessary for the _prover_ to create a zero-knowledge proof or a signature.
+
+### Commitments
+
+[Pedersen commitments](zkvm-spec.md#pedersen-commitment) are represented as enums:
+
+* `Commitment::Closed` is an _opaque type_ holding a compressed Ristretto [point](zkvm-spec.md#point) (represented by a 32-byte string).
+* `Commitment::Open ` is a _witness type_ that holds a pair of a secret value ([scalar witness](#scalar-witness)) and a secret blinding factor. These are used to create a R1CS proof using the prover’s instance of the VM.
+
+### Predicates
+
+`Predicate` type represents either an opaque [predicate](zkvm-spec.md#predicate), or an entire tree where unused branches can be left opaque and visited branches are stored as-is. This structure can then produce correct input for the [`call`](zkvm-spec.md#call), [`left`](zkvm-spec.md#left) and [`right`](zkvm-spec.md#right) instructions.
+
+### Variables
+
+Variables are represented as indices into the commitments stored within the VM state. VM manages the "attached" state of the variables to permit replacing the commitments via [`reblind`](zkvm-spec.md#reblind). The secret assignments for the variables are stored within their commitments as [described above](#commitments).
+
+### Expressions
+
+Expressions store their witness data as an optional assignment `Option<ScalarWitness>`.
+The verifier’s VM sees `None` and the prover’s VM sees `Some(...)`.
+
+See also [scalar witness](#scalar-witness).
+
+### Constraints
+
+Constraints do not have explicit witness data. It can be computed on the fly by evaluating the boolean function that the constraint represents, taking the underlying [expressions’](#expressions) assignments as input.
```

Comment:
(nit): It -> They
(since "Constraints" is plural)

Reply:
thx